pkg/resource/policy/hooks.go

```diff
@@ -191,3 +221,242 @@ func (rm *resourceManager) removeTags(
+		err = rm.deletePolicyVersion(ctx, policyARN, versions[0].version)
+		if err != nil {
+			return err
+		}
+		rlog.Info(
+			"exceeded limit of policy versions. deleted earliest policy "+
+				"version before adding new version.",
+			"policy_version", versions[0].version,
+		)
+	}
+
+	return nil
+}
+
+type policyVersion struct {
+	version    string
+	createDate *time.Time
+	document   string
+	isDefault  bool
+}
+
+// getPolicyVersion gets the specified policy version from the supplied policy
+func (rm *resourceManager) getPolicyVersion(
+	ctx context.Context,
+	policyARN string,
+	version string,
+) (pv *policyVersion, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.getPolicyVersion")
+	defer func(err error) {
+		exit(err)
+	}(err)
+
+	input := &svcsdk.GetPolicyVersionInput{}
+	input.SetPolicyArn(policyARN)
+	input.SetVersionId(version)
+
+	var resp *svcsdk.GetPolicyVersionOutput
+	resp, err = rm.sdkapi.GetPolicyVersionWithContext(ctx, input)
+	rm.metrics.RecordAPICall("READ_ONE", "GetPolicyVersion", err)
+
+	if err != nil {
+		return nil, err
+	}
+	pv = &policyVersion{}
+	if resp.PolicyVersion != nil {
+		pv.createDate = resp.PolicyVersion.CreateDate
+		if resp.PolicyVersion.VersionId != nil {
+			pv.version = *resp.PolicyVersion.VersionId
+		}
+		if resp.PolicyVersion.Document != nil {
+			// The policy document is URL-encoded by default, which leads to
+			// false positive deltas...
+			rawDoc := *resp.PolicyVersion.Document
+			doc, err := url.QueryUnescape(rawDoc)
+			if err != nil {
+				return nil, err
+			}
+			pv.document = doc
+		}
+		if resp.PolicyVersion.IsDefaultVersion != nil {
+			pv.isDefault = *resp.PolicyVersion.IsDefaultVersion
+		}
+	}
+	return pv, nil
+}
+
+// getPolicyVersions returns a slice, sorted by creation date, of the policy
+// versions for a supplied Policy ARN.
+func (rm *resourceManager) getPolicyVersions(
+	ctx context.Context,
+	policyARN string,
+) (versions []policyVersion, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.getPolicyVersions")
+	defer func(err error) {
+		exit(err)
+	}(err)
+
+	input := &svcsdk.ListPolicyVersionsInput{}
+	input.SetPolicyArn(policyARN)
+
+	var resp *svcsdk.ListPolicyVersionsOutput
+	versions = []policyVersion{}
+	for {
+		resp, err = rm.sdkapi.ListPolicyVersionsWithContext(ctx, input)
+		if err != nil || resp == nil {
+			break
+		}
+		for _, v := range resp.Versions {
+			// NOTE(jaypipes): Deliberately skipping the PolicyDocument because
+			// we don't use this information in callers of this function. The
+			// singular getPolicyVersion() method *does* return the
+			// PolicyDocument, however, and that method is called to populate
+			// the Spec.PolicyDocument in sdkFind()
+			pv := policyVersion{
+				version:    *v.VersionId,
+				createDate: v.CreateDate,
+				isDefault:  *v.IsDefaultVersion,
+			}
+			versions = append(versions, pv)
+		}
+		if resp.IsTruncated != nil && !*resp.IsTruncated {
+			break
+		}
```

Comment:
Missing pagination `Marker` setter here.
However i don't understand why there is a pagination mechanism if you can only have 5 policy versions at most..

```suggestion
		if resp.IsTruncated != nil && !*resp.IsTruncated {
			break
		}
		input.SetMarker(*resp.Marker)
                rm.metrics.RecordAPICall("READ_MANY", "ListPolicyVersions", err)
```

Reply:
ooh, nice catch, I always seem to forget this. Not urgent, since you actually can't have more than 5 policy versions, but still needed. I will fix this for tags as well.